lemur/plugins/lemur_acme/challenge_types.py

```diff
@@ -0,0 +1,240 @@
+
+        :param csr:
+        :param issuer_options:
+        :return:
+        """
+        pass
+
+    def deploy(self, challenge, acme_client, validation_target):
+        """
+        In here the challenge validation is fetched and deployed somewhere that it can be validated by the provider
+
+        :param self:
+        :param challenge: the challenge object, must match for the challenge implementation
+        :param acme_client: an already bootstrapped acme_client, to avoid passing all issuer_options and so on
+        :param validation_target: an identifier for the validation target, e.g. the name of a DNS provider
+        """
+        raise NotImplementedError
+
+    def cleanup(self, challenge, acme_client, validation_target):
+        """
+        Ideally the challenge should be cleaned up, after the validation is done
+        :param challenge: Needed to identify the challenge to be removed
+        :param acme_client: an already bootstrapped acme_client, to avoid passing all issuer_options and so on
+        :param validation_target: Needed to remove the validation
+        """
+        raise NotImplementedError
+
+
+class AcmeHttpChallenge(AcmeChallenge):
+    challengeType = challenges.HTTP01
+
+    def create_certificate(self, csr, issuer_options):
+        """
+        Creates an ACME certificate using the HTTP-01 challenge.
+
+        :param csr:
+        :param issuer_options:
+        :return: :raise Exception:
+        """
+        self.acme = AcmeHandler()
+        authority = issuer_options.get("authority")
+        acme_client, registration = self.acme.setup_acme_client(authority)
+
+        orderr = acme_client.new_order(csr)
+
+        chall = []
+        for authz in orderr.authorizations:
+            # Choosing challenge.
+            # authz.body.challenges is a set of ChallengeBody objects.
+            for i in authz.body.challenges:
+                # Find the supported challenge.
+                if isinstance(i.chall, challenges.HTTP01):
+                    chall.append(i)
+
+        if len(chall) == 0:
+            raise Exception('HTTP-01 challenge was not offered by the CA server.')
+        else:
+            validation_target = None
+            for option in json.loads(issuer_options["authority"].options):
+                if option["name"] == "tokenDestination":
+                    validation_target = option["value"]
+
+            if validation_target is None:
+                raise Exception('No token_destination configured for this authority. Cant complete HTTP-01 challenge')
+
+        for challenge in chall:
+            response = self.deploy(challenge, acme_client, validation_target)
+
+            acme_client.answer_challenge(challenge, response)
+
+        current_app.logger.info("Uploaded HTTP-01 challenge tokens, trying to poll and finalize the order")
+
+        finalized_orderr = acme_client.finalize_order(orderr, datetime.datetime.now() + datetime.timedelta(seconds=90))
```

Comment:
I assumed, that the file upload is done within the 90 second timeout defined in `finalize_order`, and because right now, the cert bootstrap isn't done asynchronously I didn't add much error handling.

But it would definitely be better to add a nice and readable error message that is displayed to the user, instead of the ACME errors

Reply:
I've added quite a bit more exception handling to the http challenge, as well as the upload function of the sftp destination